SU2_CFD/src/solvers/CGradientSmoothingSolver.cpp

```diff
@@ -0,0 +1,1168 @@
+      LinSysSol.Initialize(nPoint, nPointDomain, 1, 0.0);
+      LinSysRes.Initialize(nPoint, nPointDomain, 1, 0.0);
+      Jacobian.Initialize(nPoint, nPointDomain, 1, 1, false, geometry, config, false, true);
+    }
+  }
+
+  /*--- initializations for the debug mode ---*/
+  if (config->GetSobMode()==DEBUG) {
+    auxVec.Initialize(nPoint, nPointDomain, nDim, 1.0);
+  }
+
+  /*--- vectors needed for projection when working on the complete system ---*/
+  activeCoord.Initialize(nPoint, nPointDomain, nDim, 0.0);
+  helperVecIn.Initialize(nPoint, nPointDomain, nDim, 0.0);
+  helperVecOut.Initialize(nPoint, nPointDomain, nDim, 0.0);
+  helperVecAux.Initialize(nPoint, nPointDomain, nDim, 0.0);
+
+  /*--- Initialize the CVariable structure holding solution data ---*/
+  nodes = new CSobolevSmoothingVariable(nPoint, nDim,  config);
+  SetBaseClassPointerToNodes();
+
+  /*--- Initialize the boundary of the boundary ---*/
+  if (config->GetSmoothOnSurface()) {
+
+    /*--- check which points are in more than one physical boundary ---*/
+    for (iPoint = 0; iPoint < nPoint; iPoint++) {
+      for (unsigned int iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++) {
+        if (config->GetMarker_All_KindBC(iMarker) != SEND_RECEIVE) {
+          long iVertex = geometry->nodes->GetVertex(iPoint, iMarker);
+          if (iVertex >= 0) {
+            marker_count++;
+          }
+        }
+      }
+      if (marker_count>=2) {
+        nodes->MarkAsBoundaryPoint(iPoint);
+      }
+      marker_count = 0;
+    }
+  }
+
+  /*--- Initialize ij-block for the Jacobian ---*/
+  Jacobian_block = new su2double*[nDim];
+  for (iDim = 0; iDim < nDim; iDim++) {
+    Jacobian_block[iDim] = new su2double [nDim];
+    for (jDim = 0; jDim < nDim; jDim++) {
+      Jacobian_block[iDim][jDim] = 0.0;
+    }
+  }
+
+  /*--- vector for the parameter gradient ---*/
+  for (auto iDV=0; iDV<config->GetnDV_Total(); iDV++) {
+    deltaP.push_back(0.0);
+  }
+
+  /*--- set the solver name for output purposes ---*/
+  SolverName = "SOBOLEV";
+}
+
+CGradientSmoothingSolver::~CGradientSmoothingSolver(void) {
+
+  unsigned int iDim;
+
+  if (element_container != nullptr) {
+    for (unsigned int iVar = 0; iVar < MAX_TERMS; iVar++) {
+      for (unsigned int jVar = 0; jVar < MAX_FE_KINDS; jVar++) {
+        delete element_container[iVar][jVar];
+      }
+      delete [] element_container[iVar];
+    }
+    delete [] element_container;
+  }
+
+  if (Jacobian_block != nullptr) {
+    for (iDim = 0; iDim < nDim; ++iDim) {
+      delete [] Jacobian_block[iDim];
+    }
+    delete [] Jacobian_block;
+  }
+
+  for (iDim = 0; iDim < nDim; iDim++) {
+    if (mId_Aux[iDim] != nullptr) delete [] mId_Aux[iDim];
+  }
+  if (mId_Aux != nullptr) delete [] mId_Aux;
+
+  delete nodes;
+
+}
+
+void CGradientSmoothingSolver::ApplyGradientSmoothingVolume(CGeometry *geometry, CNumerics *numerics, CConfig *config) {
+
+  /*--- current dimension if we run consecutive on each dimension ---*/
+  dir = 0;
```

Comment:
local variable if you don't use it elsewhere.
The "guideline" is to reduce the state of the class to the minimum essential.

Reply:
This is needed in several functions. 
Since they are overloads of existing CSolver virtual routines I do not want to add an additional variable to them. I switched it so that the variable dir in the solver is now accessed by Set and Get functions for more clarity.